src/main/java/com/uber/jenkins/phabricator/coverage/CoberturaXMLParser.java

```diff
@@ -118,25 +118,34 @@ public InputSource resolveEntity(String publicId, String systemId) throws SAXExc
             NodeList classes = entry.getKey();
             List<String> sourceDirs = entry.getValue();
 
-            // Loop over all files in the coverage report
+            // Collect all filenames in coverage report
+            List<String> fileNames = new ArrayList<String>();
+            List<NodeList> childNodes = new ArrayList<NodeList>();
             for (int i = 0; i < classes.getLength(); i++) {
                 Node classNode = classes.item(i);
                 String fileName = classNode.getAttributes().getNamedItem(NODE_FILENAME).getTextContent();
 
                 if (includeFileNames != null && !includeFileNames.contains(FilenameUtils.getName(fileName))) {
                     continue;
                 }
+                fileNames.add(fileName);
+                childNodes.add(classNode.getChildNodes());
+            }
+
+            // Make multiple guesses on which of the `sourceDirs` contains files in question
+            Map<String, String> detectedSourceRoots = new PathResolver(workspace, sourceDirs).choose(fileNames);
 
-                // Make a guess on which of the `sourceDirs` contains the file in question
-                String detectedSourceRoot = new PathResolver(workspace, sourceDirs).choose(fileName);
-                fileName = join(detectedSourceRoot, fileName);
+            // Loop over all files which are needed for coverage report
+            for (int i = 0; i < fileNames.size(); i++) {
```

Comment:
@kageiit - now I'm not looping again through same loop although it requires us to collect childNodes for this loop to work